Source/CarthageKit/Project.swift

```diff
@@ -540,38 +540,57 @@ public final class Project {
 					.map { (cartfile: Cartfile) -> DependencyGraph in
 						return [ dependency.project: Set(cartfile.dependencies.map { $0.project }) ]
 					}
+					.concat(SignalProducer(value: [ dependency.project: Set() ]))
+					.take(1)
 			}
-			.collect()
-			.flatMap(.Latest) { (graphs: [DependencyGraph]) -> SignalProducer<Dependency<PinnedVersion>, CarthageError> in
-				let graph = graphs.reduce([:]) { (graph: DependencyGraph, item: DependencyGraph) -> DependencyGraph in
-					var result = graph
-					for (dependency, projects) in item {
-						var recursive = projects
-						for project in projects {
-							if let projects = result[project] {
-								recursive.unionInPlace(projects)
-							}
-						}
-						
-						result.updateValue(recursive, forKey: dependency)
-						
-						for (project, projects) in result {
-							if projects.contains(dependency) {
-								result.updateValue(projects.union(recursive), forKey: project)
-							}
+			.reduce([:]) { (working: DependencyGraph, next: DependencyGraph) in
+				var result = working
+				next.forEach { result.updateValue($1, forKey: $0) }
+				return result
+			}
+			.flatMap(.Latest) { (graph: DependencyGraph) -> SignalProducer<Dependency<PinnedVersion>, CarthageError> in
+				// Open Question: Is there a place that a generic topological 
+				// sort should live?
+				func topologicalSort<Node: Equatable>(edges: Dictionary<Node, Set<Node>>, @noescape isOrderedBefore: (Node, Node) -> Bool) -> [Node] {
+					var nodeQueue: [Node] = edges.filter { $1.isEmpty }.map { return $0.0 }
+
+					var workingEdges = edges
+					nodeQueue.forEach { workingEdges.removeValueForKey($0) }
+
+					var sortedNodes: [Node] = []
+
+					while !nodeQueue.isEmpty {
+						nodeQueue.sortInPlace(isOrderedBefore)
+
+						let lastNode = nodeQueue.removeLast()
+						sortedNodes.append(lastNode)
+
+						for (node, nodeEdges) in workingEdges {
+							guard nodeEdges.contains(lastNode) else { continue }
+
+							let unsharedEdges = nodeEdges.subtract([lastNode])
+							workingEdges[node] = unsharedEdges
+
+							guard unsharedEdges.isEmpty else { continue }
+							nodeQueue.append(node)
 						}
 					}
-					return result
+
+					// There is a cycle in the graph. How should we communicate
+					// this?
+//					if sortedNodes.count != edges.count {
+//
+//					}
```

Comment:
Since it's the only way that this can fail, I think returning `[Node]?` would work. So long as we document that it's what `nil` means.

But `buildOrderForResolvedCartfile` should return a suitable `CarthageError`. It doesn't look like we have one now for a build cycle, so one would need to be added.